decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,600 @@
+  Geometry <-- Sequence_3
+  Geometry <-- Rig
+  Blob <-- Image
+  Sequence_1 <-- RenderCache
+  Image <-- Sequence_1
+  Image <-- Panoramic
+  Sequence_2 <-- Shot
+  Container <-- Sequence_2
+  Sequence_3 <-- AnimationCache
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence_1 {
+    int startFrame
+    int endFrame
+  }
+  class Sequence_2 {
+    int startFrame
+    int endFrame
+  }
+  class Sequence_3 {
+    int startFrame
+    int endFrame
+  }
+  
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+#### Option 2
+
+This lets us specify only the properties that apply to the
+specification, without duplication.
+
+```mermaid
+classDiagram
+  direction RL
+  Rig *-- Geometry
+  Rig *-- Blob
+  
+  AnimationCache *-- Blob
+  AnimationCache *-- Geometry
+  AnimationCache *-- Sequence
+  AnimationCache *-- Cache
+  RenderCache *-- Image
+  RenderCache *-- Blob
+  RenderCache *-- Sequence
+  RenderCache *-- Cache
+  Panoramic *-- Image
+  Panoramic *-- Blob
+  Panoramic *-- Lens
+  Shot *-- Container
+  Shot *-- Sequence
+  class Blob {
+    str mimeType
+  }
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+  }
+  class Rig {
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+  }
+  class Panoramic {
+  }
+  class Lens {
+    int angle
+  }
+  class Sequence {
+    int startFrame
+    int endFrame
+  }
+  class Cache {
+    int size
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+```
+  trait Blob {
+    str mimeType
+  }
+  trait Geometry {
+    str upAxis
+  }
+  trait Image {
+    int channels
+  }
+  trait Lens {
+    int angle
+  }
+  trait Sequence {
+    int startFrame
+    int endFrame
+  }
+  trait Cache {
+    int size
+  }
+  trait Container {
+    bool isEmpty
+  }
+  
+  specification AnimationCache {
+    traits: blob, geometry, sequence, cache
+  }
+  specification Rig {
+    traits: blob, geometry
+  }
+  specification RenderCache {
+    traits: blob, sequence, image, cache
+  }
+  specification Panoramic {
+    traits: blob, image, lens
+  }
+```
+
+### Checking management policies
+
+In this case, for Option 1 and 2 since blob is part of the shared
+hierarchy for the requested specifications in Option 1, and a trait in
+Option 2, the usage is very similar.
+
+#### Option 1
+
+When using management interface functions the code using specifications
+is the same for Option 1 and Option 2.
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+policy = manager.managementPolicy([ImageSpecification()], context)
+
+if policy == [kIgnored]:
+    pass  # nothing with an ImageSpecification is managed
+
+# check if pngs and alembic caches are supported by the manager
+exrSpec = BlobSpecification()
```

Comment:
[question] Is there a reason for moving away from using `Blob.Image` here, and in the rest of the examples? The idea of specifications is that they should be used as precisely as possible, in general the policy would be more subject to the use of a specific type of data (eg: renders vs shader textures), over the format itself (eg: png).

Reply:
It's to show that they can be done similarly in some cases. BlobSpecification is pretty similar to BlobTrait in the case that the hierarchy is the same as the shared traits.

Reply:
Capturing an off-line conversation:

> Properties of specs can be a slightly grey area. The general rule is that the schema/traits should be enough to concretely define something. Ie: `rig.hair` and `rig.mesh` rather than `rig` with a `department` property, so it's best to use the most concrete definition at all times.

Reply:
No problem, here it is: 70bea88

Reply:
How would it be updated? I can't find the line.

Reply:
`L464` says:

> When using management interface functions the code using specifications is the same for Option 1 and Option 2.

The code for option 2 at `L496` and `L499` still uses a base `BlobTrait`, by the "most concrete definition" principal it should also be using an image or geometry specification so that it is "the same" as the other option (at that point do we just need a single code example?).

if you want this to be a realistic example (and avoid the property debate) I'd almost suggest not setting a `mimeType` on at least the image example. In all the situations we've implemented in the past, its rare that a specific type would be queried. Taking Nuke for example, to query if image read nodes should even bother the AMS (a concrete example of where `managementPolicy` should be used), would not specify a mime type as the type of the image could me multitude, and most likely determined by the pipeline, not the host.

Reply:
The point was more that they _should_ be the same code. The use of `BlobTrait` as it is presented for Option 2 is not in the spirit of how specifications should be used.